### DISCUSSION TITLE
Feature request: support for other modules with TFT/OLED

### DISCUSSION THREAD

**Discussion by mrnorberthorvath** [PlummersSoftwareLLC/NightDriverStrip#33] [Ideas]
Support for some other common ESP32 modules with screens would be very nice, for example the Lilygo/TT-GO T-display ESP32 board has an ST7789 display if I am right. Sadly I haven't watched enough Youtube tutorials yet to be able to give an acceptable solution.

**Comment by BeldrothTheGold** [comment]
> Support for some other common ESP32 modules with screens would be very nice, for example the Lilygo/TT-GO T-display ESP32 board has an ST7789 display if I am right. Sadly I haven't watched enough YouTube tutorials yet to be able to give an acceptable solution.

Technically, the M5Stick, for which the spectrum example was written, uses the same ST7789 display.

We could see if Dave would be willing to use a generic ST7789 driver library vs using the M5 provided abstraction. It would probably make it easier to support other ESP32 with the same kind of display, like you suggested. 

--

On a side note, the Heltec actually has a monochrome OLED and the M5Stick has a color TFT.
So the USE_TFT define tells the tool to enable the SSD1306 OLED display on the Heltec and the USE_OLED define tells the tool to enable the TFT on the M5Stick.  😄

**Comment by robertlipe** [comment]
This question is a bit of a "how long is a rope?" question. No matter how many ST7735/7739/GC9203/ILI9341/ILI9163/Hercules MDA/e-paper/telegraph/punched card/smoke signal interfaces you support, someone will always make a case for yet another. That's not terrible; it's just open-ended.

The good news, in all the years this question has been lingering without the "answered" button being checked, is that both M5Stick and various Lilygo products have had support fleshed out. (Before you ask, I have none of these...I'm familiar with the problem space and am "just" trying to help cleanup the backlog of open questions.) 

ledstrip_feather, for example, seems to be based on Adafruit's popular Feather design , using their TFTSPI code for ST7789 and plumbs up all the pinning:

```
[env:ledstrip_feather]
extends         = dev_adafruit_feather
build_flags     = -DLEDSTRIP=1
                  -DUSE_SCREEN=1
                  -DUSE_TFTSPI=1
                  -DUSER_SETUP_LOADED
                  -DTOUCH_CS=0
                  -DST7789_2_DRIVER
                  -DTFT_WIDTH=135                   ; Define the TFT screen setup
                  -DTFT_HEIGHT=240
                  -DTFT_BL=45
                  -DTFT_BACKLIGHT_ON=HIGH
                  -DTFT_RGB_ORDER=TFT_RGB           ; Fix color order being swapped
                  -DTFT_CS=7                        ; CS, DC, RST
                  -DTFT_DC=39
                  -DTFT_RST=40
                  -DTFT_MOSI=35                     ; MOSI, MISO, CLK
                  -DTFT_MISO=37
                  -DTFT_SCLK=36
```

Other combinations in plat*ini trigger USE_SCREEN. Notably the M5STICKCPLUS, _m5stick-c-plus2, m5stack, and others seem to all light up code in the M5 library. For Lilygo, there are a number of examples like:
```
[dev_lilygo_amoled]

extends        = base
board          = lilygo-t-amoled
monitor_speed  = 115200
upload_speed   = 1500000
board_build.partitions = config/partitions_custom_noota.csv
build_flags    = ${base.build_flags}
                 -DLILYGO_TDISPLAY_AMOLED_SERIES=1
                 -DARDUINO_USB_CDC_ON_BOOT=1
                 -DAMOLED_S3=1
                 -DUSE_SCREEN=1
                 -DTOGGLE_BUTTON_1=21
                 -DTOGGLE_BUTTON_2=0
                 -DNUM_INFO_PAGES=2
                 -DLED_PIN0=2
                 -DLV_CONF_PATH="../../../../include/amoled/lv_conf.h"
                 -DLV_CONF_INCLUDE_SIMPLE
```

or ledstrip_feather_hexagon, spectrum_wrover_kit, panlee, ttgo, atomlight, heltec_wifi], m5stick-(c,plus,plus2) and more that set up complete examples of these custom baked configurations.

heltecv3demo uses 1306.

There's a whole layer for:
```
src/screen.cpp
25://    Handles the TFT or OLED display that is optionally connected.  It
42:#if USE_TFTSPI
43:#include <TFT_eSPI.h>
```
and the eSPI code supports a brazillion panels and controllers if you "just" build up the right pin and initialization configurations.

So I don't know how bad the situation was in 2021 when this question was last touched, but here in 2024, the number of examples of "how can I support display X on hardware Y with effect combination Z" is not a fully closed set, but it's a large amount of code and examples from which to pull inspiration. There are a LOT of TFT/OLED displays supported in some form these days.

IMO, the general "support for other modules" case is now covered and should be marked as answered.

If there's a specific remaining "I need help supporting X on Y with Z", that's probably best treated as an individual consulting request, perhaps best accelerated by sending  representative hardware, precise requirements, and motivation (e.g. preferred beverages) to build up the precise remaining custom configuration and hopefully help get that integrated as an example for others.
